### PR TITLE
version one of lsm system

### PR DESCRIPTION
**Pull request by mmfrenkel**
Thank you for your review! The two-three tree is not supported/used by current iteration.

### CODE REVIEW

File: lsm_tree.c [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+		if (value != NULL) {
+			printf("The value for key %d is %s.\n", submission->key, value);
+		} else {
+			printf("Key not found in LSM Tree system.\n");
+		}
+
+	} else if (submission->action == DELETE) {
+		// always soft delete here; do not decrement keys in tree
+		if (delete(lsm_tree->memtable, submission->key, false) != 0) {
+			printf("Deletion of key, value pair with key %d failed.\n",
```
Comment (andyg7) [inline lsm_tree.c]: np; for deletion the value should just be null so probably can just comment something along lines of ```Deletion of key failed```

File: lsm_tree.c [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+	} else if (submission->action == PRINT_MEMTABLE) {
+		print_memtable(lsm_tree->memtable, "in_order_traversal");
+
+	} else {
+		printf("Could not process request in submission.\n");
+		return -1;
+	}
+
+	// this should be in the background, rather than sequential in future
+	return run_compaction(lsm_tree);
```
Comment (andyg7) [inline lsm_tree.c]: How about maintaining a counter and running the compaction only every X operations?

File: lsm_tree.c [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+				fputs("\n", new_fp);
+			} while (fgets(temp_line, MAX_LINE_SIZE, temp_ptr) != NULL);
+
+			keep_merging = false;
+
+		} else
+			merge_lines(line_a, line_b, new_fp, &incr_ptr_a, &incr_ptr_b);
+	}
+	// close file pointers for new file and the two segment files
+	fclose(new_fp);
```
Comment (andyg7) [inline lsm_tree.c]: Let's handle the errors here? Same for other uses of ```fclose```

File: lsm_tree.c (hunk 2 above)
Comment (andyg7) [inline lsm_tree.c]: How about maintaining a counter and only compacting every X operations?

File: lsm_tree.c (hunk 1 above)
Comment (andyg7) [inline lsm_tree.c]: np; because deletion shouldn't have a value could just print ```Deletion of key %d failed```

File: lsm_tree.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+#define STR_BUF 5              // leave plenty of room for options
+#define MAX_LEN_KEYS 10        // max size of key (10 digits)
+#define MAX_LEN_DATA 50        // max length of data for value in database
+#define MAX_KEYS_IN_TREE 3     // max # keys held in tree before flush to segment
+#define MAX_SEGMENTS 2         // max # full segments before compaction
+#define FILENAME_SIZE 20       // file name size
+#define MAX_LINE_SIZE 60       // max number of characters in a single line of a segment file
+
+enum available_actions {
+	ADD = 1, SEARCH = 2, DELETE = 3, SAVE = 4, PRINT_MEMTABLE = 5, EXIT = 6
```
Comment (andyg7) [inline lsm_tree.h]: np; instead of ```SAVE``` I think ```FLUSH``` might be more idiomatic

File: lsm_tree.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+};
+
+typedef struct user_submission {
+	enum available_actions action;
+	int key;
```
Comment (andyg7) [inline lsm_tree.h:21]: Just curious why is ```key``` an ```int```? Personal preference would be a ```char *```

File: lsm_tree.c [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+	fclose(seg_ptr_a);
+	fclose(seg_ptr_b);
+	return new_segment;
+}
+
+/* Helper function for merging two files together by adding lines from either file a
+ * or file b to the new file, then assigning values to determine whether the
+ * calling function should increment a or b */
+static void merge_lines(char *line_a, char *line_b, FILE *new_fp, bool *incr_a,
+bool *incr_b) {
```
Comment (andyg7) [inline lsm_tree.c]: np; move up a line

File: lsm_tree.c [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+	lsm_tree->memtable = memtable;
+	lsm_tree->segments = segments;
+	lsm_tree->full_segments = 0;
+	return lsm_tree;
+}
+
+/* Handles the submission provided by user; returns 0 if all succeeds,
+ * otherwise returns -1.  */
+int handle_submission(LSM_Tree *lsm_tree, Submission *submission) {
+	if (submission->action == ADD) {
```
Comment (andyg7) [inline lsm_tree.c:84]: For each action might want to add some small validation. Off the top of my head I'm thinking for ```ADD``` validate the value is not the tombstone ```string``` and for ```DELETE``` validate value is ```null```

File: custom_io.c [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+#define FILENAME_SIZE 20
+
+/* prototypes for static functions */
+static void serialize_preorder(MNode *root, FILE *fp);
+static void add_to_file(char *filename, char *to_write);
+static void inorder_to_file(MNode *root, FILE *fp);
+
+/* Writes a memtable (binary tree) to file */
+void serialize_memtable(Memtable *memtable, char *filename) {
+	FILE *fp;
```
Comment (andyg7) [inline custom_io.c:24]: np; could be one line ```FILE *fp =  fopen(filename, "w");``` Also, can ```fopen``` return an error?

**Review by andyg7 — COMMENTED** [review]
File structure of the project looks great.

File: lsm_tree.c (hunk 1 above)
Comment (mmfrenkel) [reply lsm_tree.c]: Changed wording to only mention key.

File: lsm_tree.c (hunk 3 above)
Comment (mmfrenkel) [reply lsm_tree.c]: ya, you're right I should check. I added error checking for the important cases.

File: lsm_tree.c (hunk 2 above)
Comment (mmfrenkel) [reply lsm_tree.c]: refactored to make it clearer that compaction should only run on specific occasions.

Comment (mmfrenkel) [reply lsm_tree.c]: repeat comment -- resolving

File: lsm_tree.c (hunk 1 above)
Comment (mmfrenkel) [reply lsm_tree.c]: repeat comment -- resolving

File: lsm_tree.h (hunk 4 above)
Comment (mmfrenkel) [reply lsm_tree.h]: okay!

File: lsm_tree.h (hunk 5 above)
Comment (mmfrenkel) [reply lsm_tree.h:21]: Yea, that's a good point. It probably should be a char*. Using a key is easy for making comparisons and order which was why I decided to go with that first. Will change in a next iteration.

File: lsm_tree.c (hunk 7 above)
Comment (mmfrenkel) [reply lsm_tree.c:84]: Not sure what you mean the delete step, but changed the `ADD` to check for tombstone.

File: lsm_tree.c (hunk 6 above)
Comment (mmfrenkel) [reply lsm_tree.c]: fixed!

File: custom_io.c (hunk 8 above)
Comment (mmfrenkel) [reply custom_io.c:24]: fixed!

**Pull request merged**